lib/onelogin/ruby-saml/response.rb

```diff
@@ -228,20 +546,100 @@ def validate_conditions(soft = true)
 
+      # Validates that the Session haven't expired (If the response was initialized with the :allowed_clock_drift option,
+      # this time validation is relaxed by the allowed_clock_drift value)
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if the SessionNotOnOrAfter of the AttributeStatement is valid, otherwise (when expired) False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_session_expiration(soft = true)
+        return true if session_expires_at.nil?
+
+        now = Time.now.utc
+        unless session_expires_at > (now + (options[:allowed_clock_drift] || 0))
+          error_msg = "The attributes have expired, based on the SessionNotOnOrAfter of the AttributeStatement of this Response"
+          @errors << error_msg
+          return soft ? false : validation_error(error_msg)
+        end
+
+        true
+      end
+
+      # Validates if exists valid SubjectConfirmation (If the response was initialized with the :allowed_clock_drift option,
+      # timimg validation are relaxed by the allowed_clock_drift value)
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if exists a valid SubjectConfirmation, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_subject_confirmation(soft = true)
+        valid_subject_confirmation = false
+
+        subject_confirmation_nodes = xpath_from_signed_assertion('/a:Subject/a:SubjectConfirmation')
+        
+        now = Time.now.utc
+        subject_confirmation_nodes.each do |subject_confirmation|
+          if subject_confirmation.attributes.include? "Method" and subject_confirmation.attributes['Method'] != 'urn:oasis:names:tc:SAML:2.0:cm:bearer'
+            next
+          end
+
+          confirmation_data_node = REXML::XPath.first(subject_confirmation, 'a:SubjectConfirmationData', { "a" => ASSERTION })
+
+          if not confirmation_data_node
+            next
+          else
+            attrs = confirmation_data_node.attributes
+            next if (attrs.include? "InResponseTo" and attrs['InResponseTo'] != in_response_to) ||
+                    (attrs.include? "Recipient" and attrs['Recipient'] != current_url) ||
+                    (attrs.include? "NotOnOrAfter" and (parse_time(confirmation_data_node, "NotOnOrAfter") + (options[:allowed_clock_drift] || 0)) <= now) ||
+                    (attrs.include? "NotBefore" and parse_time(confirmation_data_node, "NotBefore") > (now + (options[:allowed_clock_drift] || 0)))
+            
+            valid_subject_confirmation = true
+            break
```

Comment:
If we find a valid_subject_confirmation, then we don't need to loop more the subject_confirmation_nodes